src/libPMacc/include/functor/Filtered.hpp

```diff
@@ -0,0 +1,255 @@
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * and the GNU Lesser General Public License along with libPMacc.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "pmacc_types.hpp"
+#include "filter/Interface.hpp"
+#include "mappings/threads/WorkerCfg.hpp"
+
+#include <string>
+
+
+namespace PMacc
+{
+namespace functor
+{
+namespace acc
+{
+
+
+    /** interface to combine a filter and a functor on the accelerator
+     *
+     *
+     * @tparam T_FilterOperator PMacc::filter::operators, type concatenate the
+     *                          results of the filter
+     * @tparam T_Filter PMacc::filter::Interface, type of the filter
+     * @tparam T_Functor PMacc::functor::Interface, type of the functor
+     */
+    template<
+        typename T_FilterOperator,
+        typename T_Filter,
+        typename T_Functor
+    >
+    struct Filtered :
+        private T_Filter,
+        private T_Functor
+    {
+        using Filter = T_Filter;
+        using Functor = T_Functor;
+
+        DINLINE Filtered(
+            Filter const & filter,
+            Functor const & functor
+        ) :
+            Filter( filter ),
+            Functor( functor )
+        {
+
+        }
+
+        /** execute the functor depending of the filter result
+         *
+         * Call the filter for each argument and if the combined result is true
+         * the user functor is called.
+         *
+         * @param args arguments passed to the functor
+         *             - the filter is executed for each argument
+         *             - the functor get passed all arguments
+         */
+        template< typename ... T_Args >
+        DINLINE auto operator( )(  T_Args && ... args )
+        -> void
+        {
+            if(
+                T_FilterOperator{ }(
+                    ( *reinterpret_cast< Filter * >( this ) )( args ) ...
+                )
+            )
+                ( *reinterpret_cast< Functor * >( this ) )( args ... );
+        }
+    };
+
+} // namespace acc
+
+    /** combine a filter and a functor
+     *
+     * Creates a functor where each argument which is passed to
+     * the accelerator instance is evaluated by the filter and if the
+     * combined result is true the functor is executed.
+     *
+     * @tparam T_FilterOperator PMacc::filter::operators, type concatenate the
+     *                          results of the filter
+     * @tparam T_Filter PMacc::filter::Interface, type of the filter
+     * @tparam T_Functor PMacc::functor::Interface, type of the functor
+     */
+    template<
+        typename T_FilterOperator,
+        typename T_Filter,
+        typename T_Functor
+    >
+    struct Filtered;
+
+    /** specialization of Filtered (with unary filter)
+     *
+     * This specialization can only used if T_Filter is of the type PMacc::filter::Interface
+     * and T_Functor is of the type PMacc::filter::Interface
```

Comment:
Add: `. An unary filters means that each argument can only pass the same filter check before its results are combined.`